### DISCUSSION TITLE
-delayload equivalent for MinGW?

### DISCUSSION THREAD

**Discussion by TheShermanTanker** [msys2/MINGW-packages#11720] [Q&A]
Visual C++ has the option to delay loading of a dll through the -delayload:xxx.dll flag, from what I see the ucrt port should also be able to do so (Since it has the delayimp library included in its libs folder, which hints that it _can_ delay loading a dll at runtime) but I haven't been able to figure out how to do this so far, is there a flag that I haven't been able to find controlling delayed dll loading with gcc?

(On a side note I have 2 small questions: What's the equivalent of `-z origin` with this compiler? Also, -fPIC seems to have an effect when passed with this compiler when compiling for Windows, I thought it was a no-op flag for MinGW?)

**Comment by mati865** [comment]
Binutils `dlltool` has `--output-delaylib` flag that creates "delay-import library". Unfortunately docs are very limited on this topic unfortunately.

When using LLD however there is MSVC-like `--delayload=<value>` flag.

**Reply by mlt** [reply]
I don't think there is much else to document. As it is, one has to re-create an import library. FWIW [here is what I proposed for another project](https://github.com/mlt/allegro5/commit/089574b6a5b3efce08ebee3a2232cf3dd7a80f03). Essentially a wrapper for dlltool and cmake helper. But it would be nice if those were pre-packaged.
I think it would be nice to standardize the naming and encourage packagers to supply delay load import libraries for things that are built in but may not be used often. I've been thinking about delay loading AWS libraries for [hdf5 package](https://github.com/msys2/MINGW-packages/pull/28233). I supposed it is more common to use local HDF5 files. Perhaps, that is one of the reasons users are tempted to use static build not to drag unused dependency DLLs around.
